core/tee/tadb.c

```diff
@@ -719,43 +720,58 @@ static TEE_Result ta_load(struct tee_tadb_ta_read *ta)
 	return res;
 }
 
-TEE_Result tee_tadb_ta_read(struct tee_tadb_ta_read *ta, void *buf, size_t *len)
+TEE_Result tee_tadb_ta_read(struct tee_tadb_ta_read *ta, void *buf_core,
+			    void *buf_user, size_t *len)
 {
-	TEE_Result res;
+	TEE_Result res = TEE_SUCCESS;
 	const size_t sz = ta->entry.prop.custom_size + ta->entry.prop.bin_size;
 	size_t l = MIN(*len, sz - ta->pos);
+	size_t bb_len = MIN(1024U, l);
+	size_t num_bytes = 0;
+	size_t dst_len = 0;
+	void *dst = NULL;
+	void *bb = NULL;
 
 	res = ta_load(ta);
 	if (res)
 		return res;
 
-	if (buf) {
-		res = tadb_update_payload(ta->ctx, TEE_MODE_DECRYPT,
-					  ta->ta_buf + ta->pos, l, buf);
-		if (res)
-			return res;
+	if (buf_core) {
+		dst = buf_core;
+		dst_len = l;
 	} else {
-		size_t num_bytes = 0;
-		size_t b_size = MIN(256U, l);
-		uint8_t *b = malloc(b_size);
-
-		if (!b)
+		bb = bb_alloc(bb_len);
+		if (!bb)
 			return TEE_ERROR_OUT_OF_MEMORY;
+		dst = bb;
+		dst_len = bb_len;
+	}
 
-		while (num_bytes < l) {
-			size_t n = MIN(b_size, l - num_bytes);
+	/*
+	 * This loop will only run once if buf_core is non-NULL, but as
+	 * many times as needed if the bounce buffer bb is used. That's why
+	 * dst doesn't need to be updated in the loop.
+	 */
+	while (num_bytes < l) {
+		size_t n = MIN(dst_len, l - num_bytes);
 
-			res = tadb_update_payload(ta->ctx, TEE_MODE_DECRYPT,
-						  ta->ta_buf + ta->pos +
-							num_bytes, n, b);
-			if (res)
-				break;
-			num_bytes += n;
+		res = tadb_update_payload(ta->ctx, TEE_MODE_DECRYPT,
+					  ta->ta_buf + ta->pos + num_bytes,
+					  n, dst);
+		if (res) {
+			res = TEE_ERROR_SECURITY;
```

Comment:
why `TEE_ERROR_SECURITY` here instead of `res`?

Reply:
I'll return `res` instead, the same below.